src/cpp/parse.cpp

```diff
@@ -2102,28 +2104,204 @@ bool Parser::optAlignas(typet &cv)
-    break;
+    return true;
 
-  case TOK_IDENTIFIER:
+  case TOK_GCC_ATTRIBUTE_PACKED:
+    {
+      typet attr(ID_packed);
+      set_location(attr, tk);
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_TRANSPARENT_UNION:
+    {
+      typet attr(ID_transparent_union);
+      set_location(attr, tk);
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_VECTOR_SIZE:
+    {
+      cpp_tokent tk2, tk3;
+
+      if(lex.get_token(tk2)!='(')
+        return false;
+
+      exprt exp;
+      if(!rCommaExpression(exp))
+        return false;
+
+      if(lex.get_token(tk3)!=')')
+        return false;
+
+      vector_typet attr(nil_typet(), exp);
+      attr.add_source_location()=exp.source_location();
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_ALIGNED:
+    {
+      typet attr(ID_aligned);
+      set_location(attr, tk);
+
+      if(lex.LookAhead(0)=='(')
+      {
+        cpp_tokent tk2, tk3;
+
+        if(lex.get_token(tk2)!='(')
+          return false;
+
+        exprt exp;
+        if(!rCommaExpression(exp))
+          return false;
+
+        if(lex.get_token(tk3)!=')')
+          return false;
+
+        attr.add(ID_size, exp);
+      }
+
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_MODE:
+    {
+      cpp_tokent tk2, tk3;
+
+      if(lex.get_token(tk2)!='(')
+        return false;
+
+      irept name;
+      if(!rName(name))
+        return false;
+
+      if(lex.get_token(tk3)!=')')
+        return false;
+
+      typet attr(ID_gcc_attribute_mode);
+      set_location(attr, tk);
+      attr.set(ID_size, name.get(ID_identifier));
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_GNU_INLINE:
+    {
+      typet attr(ID_static);
+      set_location(attr, tk);
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_WEAK:
+    {
+      typet attr(ID_weak);
+      set_location(attr, tk);
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_ALIAS:
+    {
+      cpp_tokent tk2, tk3, tk4;
+
+      if(lex.get_token(tk2)!='(')
+        return false;
+
+      if(!rString(tk3))
+        return false;
+
+      if(lex.get_token(tk4)!=')')
+        return false;
+
+      typet attr(ID_alias);
+      set_location(attr, tk);
+      attr.move_to_sub(tk3.data);
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_SECTION:
+    {
+      cpp_tokent tk2, tk3, tk4;
+
+      if(lex.get_token(tk2)!='(')
+        return false;
+
+      if(!rString(tk3))
+        return false;
+
+      if(lex.get_token(tk4)!=')')
+        return false;
+
+      typet attr(ID_section);
+      set_location(attr, tk);
+      attr.move_to_sub(tk3.data);
+      merge_types(attr, t);
+      break;
+    }
+
+  case TOK_GCC_ATTRIBUTE_NORETURN:
+    {
+      typet attr(ID_noreturn);
+      set_location(attr, tk);
+      merge_types(attr, t);
+      break;
+    }
+
```

Comment:
I believe I had tried to mimic what the C front-end does. The obvious fix would be to swap the parameters here, but I'm not sure this will result in the attribute being picked up at all?

Reply:
Where should the attribute end up after type checking? At some point function_type is converted to code_type. At that stage the type qualifiers need to be put somewhere in the code_type node. I didn't see anything in the C++ front and that does that, so right now cpp_convert_plain_type fails because it cannot deal with a qualifier of a function type. I could fix it if I understood how the result of type checking the function should look like.

Reply:
The merge_types function in the C and C++ front end use a different convention for the order of arguments: in the C front end, it is merge_types(dest, src), while in C++ it is merge_types(src, dest). A possible fix would then be to change the merge_types function in C++ so that it adds the new type at the front of the subtypes (or rather anywhere but at the end). Thoughts?

Reply:
Changing the merge_types functions seems to work. Regression passes and I also have a fix for the -DDEBUG compilation failures and some of the linter issues. What is the process for gettings these fixes into the PR?

Reply:
Could you check whether you are able to push to my branch? If you aren't, you can certainly create a pull request in my repository (against that branch), and as soon as I merge it the changes will show up here.